docs/01-cosock.html

```diff
@@ -0,0 +1,280 @@
+local server = socket.tcp()
+server:bind(&quot;0.0.0.0&quot;, 9999)
+server:listen()
+print(&quot;listening&quot;, server:getsockname())
+local client = server:accept()
+while true do    
+    local request = assert(client:receive())
+    assert(request == &quot;ping&quot;)
+    print(&quot;sending pong&quot;)
+    client:send(&quot;pong\n&quot;)
+end
+</code></pre>
+<p>If you were to run first <code>lua ./server.lua</code> and then <code>client.lua</code> you should see each terminal print out
+their &quot;sending ...&quot; messages forever.</p>
+<p>Using cosock, we can actually write the same thing as a single application.</p>
+<p><span id="clientserver-example"></span></p>
+<pre><code class="language-lua">-- client_server.lua
+local cosock = require &quot;cosock&quot;
+local socket = require &quot;cosock.socket&quot;
+local ip = &quot;0.0.0.0&quot;
+local server = socket.tcp()
+--- Since the client and server are in the same application
+--- we can use an OS assigned port and share it across the
+--- two tasks, to coordinate the two tasks to start in the order
+--- we want, we can use a cosock channel to make sure both tasks
+--- have the same port number
+local port_tx, port_rx = cosock.channel.new()
+
+--- Spawn a task for handling the server side of the socket
+cosock.spawn(function()
+    server:bind(ip, 0)
+    local _ip, p = server:getsockname()
+    port_tx:send(p)
+    server:listen()
+    local client = server:accept()
+    while true do
+        local request = assert(client:receive())
+        print(string.format(&quot;receieved %q&quot;, request))
+        assert(request == &quot;ping&quot;)
+        print(&quot;sending pong&quot;)
+        client:send(&quot;pong\n&quot;)
+    end
+end, &quot;server task&quot;)
+
+--- Spawn a task for handling the client side of the socket
+cosock.spawn(function()
+    --- wait for the server to be ready.
+    local port = assert(port_rx:receive())
+    local client = socket.tcp()
+    client:connect(ip, port)
+    while true do    
+        print(&quot;sending ping&quot;)
+        client:send(&quot;ping\n&quot;)
+        local request = assert(client:receive())
+        assert(request == &quot;pong&quot;)
+    end
+end, &quot;client task&quot;)
+
+--- Finally we tell cosock to run our 2 coroutines until they are done
+--- which should be forever
+cosock.run()
+</code></pre>
+<p>Now if we run this with <code>lua ./client_server.lua</code> we should see the messages alternate.</p>
+<p>Notice that we called <a href="%7E/02-spawn.html"><code>cosock.spawn</code></a> twice, once for the server task and
+once for the client task, we are going to dig into that next. We also added a call to <code>cosock.run</code>
+at the bottom of our example, this function will run our tasks until there is no more work todo
+so it is important you don't forget it or nothing will happen.</p>
+<h2 id="getting-started"><a class="header" href="#getting-started">Getting Started</a></h2>
```

Comment:
I waffle back and forth on this since I want to open with a compelling "why" before getting into the "how". It might be worth moving Getting Started into its own chapter

Reply:
I have moved the Getting Started section into its [own page](https://github.com/cosock/cosock-book/pull/1/commits/e1b1fd8959dd9f32efc0705b914a1c68398c0b98), I am curious on your thoughts?

Reply:
I like it